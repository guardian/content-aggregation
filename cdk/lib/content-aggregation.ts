import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack, GuStringParameter } from '@guardian/cdk/lib/constructs/core';
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
import type { App } from 'aws-cdk-lib';
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { SqsSubscription } from 'aws-cdk-lib/aws-sns-subscriptions';
import { Queue } from 'aws-cdk-lib/aws-sqs';

export class ContentAggregation extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const app = 'content-aggregation';

		const paramPrefix = `${this.stage}/${this.stack}/${app}/`;
		const apiKey = new GuStringParameter(this, 'api-key-param', {
			fromSSM: true,
			default: `${paramPrefix}content-api-key`,
		});
		const pageSize = new GuStringParameter(this, 'page-size', {
			fromSSM: true,
			default: `${paramPrefix}page-size`,
		});
		const apiUrl =
			this.stage === 'PROD'
				? 'https://content.guardianapis.com'
				: 'http://content.code.dev-guardianapis.com';

		const tagAggregationSnsTopic = new Topic(this, 'tag-aggregation-topic', {
			topicName: 'tag-aggregations',
		});

		const tagAggregationSqsDlq = new Queue(this, 'DLQ');
		const tagAggregationSqs = new Queue(this, 'Connection', {
			enforceSSL: true,
			deadLetterQueue: {
				queue: tagAggregationSqsDlq,
				maxReceiveCount: 1,
			},
		});
		tagAggregationSnsTopic.addSubscription(
			new SqsSubscription(tagAggregationSqs),
		);

		new GuLambdaFunction(this, 'content-aggregation-lambda', {
			app,
			handler: 'index.handler',
			functionName: `content-aggregation-${this.stage}`,
			runtime: Runtime.NODEJS_22_X,
			fileName: 'src/index.ts',
			environment: {
				API_URL: apiUrl,
				API_KEY: apiKey.valueAsString,
				PAGE_SIZE: pageSize.valueAsString,
				SNS_TOPIC_ARN: tagAggregationSnsTopic.topicArn,
			},
			initialPolicy: [
				new PolicyStatement({
					effect: Effect.ALLOW,
					actions: ['sns:Publish'],
					resources: [tagAggregationSnsTopic.topicArn],
				}),
			],
		});
	}
}
