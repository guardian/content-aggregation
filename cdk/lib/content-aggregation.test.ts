import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { ContentAggregation } from "./content-aggregation";

describe("The ContentAggregation stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new ContentAggregation(app, "ContentAggregation", { stack: "flexible", stage: "TEST" });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
