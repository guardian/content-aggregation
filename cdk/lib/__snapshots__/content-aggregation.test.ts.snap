// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ContentAggregation stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuStringParameter",
      "GuStringParameter",
      "GuDistributionBucketParameter",
      "GuLambdaFunction",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "apikeyparam": {
      "Default": "/TEST/flexible/content-aggregation/content-api-key",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "pagesize": {
      "Default": "/TEST/flexible/content-aggregation/page-size",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Connection89AD5CF5": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "RedrivePolicy": {
          "deadLetterTargetArn": {
            "Fn::GetAtt": [
              "DLQ581697C4",
              "Arn",
            ],
          },
          "maxReceiveCount": 1,
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/content-aggregation",
          },
          {
            "Key": "Stack",
            "Value": "flexible",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "ConnectionContentAggregationtagaggregationtopic1D39E3E583B82066": {
      "DependsOn": [
        "ConnectionPolicyCCB96DD5",
      ],
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "Connection89AD5CF5",
            "Arn",
          ],
        },
        "Protocol": "sqs",
        "TopicArn": {
          "Ref": "tagaggregationtopicF9E62E4D",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "ConnectionPolicyCCB96DD5": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "Connection89AD5CF5",
                  "Arn",
                ],
              },
            },
            {
              "Action": "sqs:SendMessage",
              "Condition": {
                "ArnEquals": {
                  "aws:SourceArn": {
                    "Ref": "tagaggregationtopicF9E62E4D",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "sns.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "Connection89AD5CF5",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "Connection89AD5CF5",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "DLQ581697C4": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/content-aggregation",
          },
          {
            "Key": "Stack",
            "Value": "flexible",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "contentaggregationlambda162D5E2E": {
      "DependsOn": [
        "contentaggregationlambdaServiceRoleDefaultPolicy65D952F8",
        "contentaggregationlambdaServiceRoleBB07E2C1",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "flexible/TEST/content-aggregation/content-aggregation.zip",
        },
        "Environment": {
          "Variables": {
            "API_KEY": {
              "Ref": "apikeyparam",
            },
            "API_URL": "http://content.code.dev-guardianapis.com",
            "APP": "content-aggregation",
            "PAGE_SIZE": {
              "Ref": "pagesize",
            },
            "SNS_TOPIC_ARN": {
              "Ref": "tagaggregationtopicF9E62E4D",
            },
            "STACK": "flexible",
            "STAGE": "TEST",
          },
        },
        "FunctionName": "content-aggregation-TEST",
        "Handler": "index.handler",
        "LoggingConfig": {
          "LogFormat": "JSON",
        },
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "contentaggregationlambdaServiceRoleBB07E2C1",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "content-aggregation",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/content-aggregation",
          },
          {
            "Key": "Stack",
            "Value": "flexible",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "contentaggregationlambdaServiceRoleBB07E2C1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "content-aggregation",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/content-aggregation",
          },
          {
            "Key": "Stack",
            "Value": "flexible",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "contentaggregationlambdaServiceRoleDefaultPolicy65D952F8": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "tagaggregationtopicF9E62E4D",
              },
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/flexible/TEST/content-aggregation/content-aggregation.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/flexible/content-aggregation",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/flexible/content-aggregation/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "contentaggregationlambdaServiceRoleDefaultPolicy65D952F8",
        "Roles": [
          {
            "Ref": "contentaggregationlambdaServiceRoleBB07E2C1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "tagaggregationtopicF9E62E4D": {
      "Properties": {
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/content-aggregation",
          },
          {
            "Key": "Stack",
            "Value": "flexible",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "TopicName": "tag-aggregations",
      },
      "Type": "AWS::SNS::Topic",
    },
  },
}
`;
